crates/astria-core/src/protocol/transaction/v1alpha1/action_group.rs

```diff
@@ -0,0 +1,491 @@
+        ActionGroup::BundlableGeneral(BundlableGeneral)
+    ),
+    (
+        BridgeUnlockAction,
+        ActionGroup::BundlableGeneral(BundlableGeneral)
+    ),
+    (BridgeSudoChangeAction, ActionGroup::General(General)),
+    (FeeChangeAction, ActionGroup::BundlableSudo(BundlableSudo)),
+    (
+        FeeAssetChangeAction,
+        ActionGroup::BundlableSudo(BundlableSudo)
+    ),
+    (IbcRelay, ActionGroup::BundlableGeneral(BundlableGeneral)),
+);
+
+pub trait Group {
+    fn group(&self) -> ActionGroup;
+}
+
+impl Group for Action {
+    fn group(&self) -> ActionGroup {
+        match self {
+            Action::Sequence(_) => SequenceAction::GROUP,
+            Action::Transfer(_) => TransferAction::GROUP,
+            Action::ValidatorUpdate(_) => ValidatorUpdate::GROUP,
+            Action::SudoAddressChange(_) => SudoAddressChangeAction::GROUP,
+            Action::IbcRelayerChange(_) => IbcRelayerChangeAction::GROUP,
+            Action::Ics20Withdrawal(_) => Ics20Withdrawal::GROUP,
+            Action::InitBridgeAccount(_) => InitBridgeAccountAction::GROUP,
+            Action::BridgeLock(_) => BridgeLockAction::GROUP,
+            Action::BridgeUnlock(_) => BridgeUnlockAction::GROUP,
+            Action::BridgeSudoChange(_) => BridgeSudoChangeAction::GROUP,
+            Action::FeeChange(_) => FeeChangeAction::GROUP,
+            Action::FeeAssetChange(_) => FeeAssetChangeAction::GROUP,
+            Action::Ibc(_) => IbcRelay::GROUP,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BundlableGeneral;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct General;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct BundlableSudo;
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Sudo;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ActionGroup {
+    BundlableGeneral(BundlableGeneral),
+    General(General),
+    BundlableSudo(BundlableSudo),
+    Sudo(Sudo),
+}
+
+impl fmt::Display for ActionGroup {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ActionGroup::BundlableGeneral(_) => write!(f, "BundlableGeneral"),
+            ActionGroup::General(_) => write!(f, "General"),
+            ActionGroup::BundlableSudo(_) => write!(f, "BundlableSudo"),
+            ActionGroup::Sudo(_) => write!(f, "Sudo"),
+        }
+    }
+}
+
+impl From<BundlableGeneral> for ActionGroup {
+    fn from(val: BundlableGeneral) -> ActionGroup {
+        ActionGroup::BundlableGeneral(val)
+    }
+}
+
+impl From<General> for ActionGroup {
+    fn from(val: General) -> ActionGroup {
+        ActionGroup::General(val)
+    }
+}
+
+impl From<BundlableSudo> for ActionGroup {
+    fn from(val: BundlableSudo) -> ActionGroup {
+        ActionGroup::BundlableSudo(val)
+    }
+}
+
+impl From<Sudo> for ActionGroup {
+    fn from(val: Sudo) -> ActionGroup {
+        ActionGroup::Sudo(val)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum ErrorKind {
```

Comment:
Noted further down on the custom `Display` implementation on public `Error` type: lean on the `error` attribute of the `thiserror` proc macro to generate that code for you. I.e. use struct variants, for example like so:

```rust
#[derive(Debug, thiserrror::Error)]
enum ErrorKind {
    #[error("actions of different types are not permitted; input contained `{group_one}` and `{group_two}`")]
    Mixed {
        group_one: ActionGroup,
        group_two: ActionGroup,
    }
}
```